crates/bridge-proof/protocol2/src/tx_inclusion_proof.rs

```diff
@@ -0,0 +1,186 @@
+
+            // Use the coinbase transaction (index 0) as the “base” transaction.
+            let coinbase = txs[0].clone();
+            let coinbase_proof = L1TxProof::generate(txs, 0);
+            let base_tx = L1TxWithIdProof::new(coinbase.into(), coinbase_proof);
+
+            Self {
+                base_tx,
+                witness_tx,
+            }
+        }
+    }
+
+    /// Verifies this [`L1TxWithProofBundle`] against a given [`Header`].
+    ///
+    /// - If `witness_tx` is `None`, this simply verifies that the `base_tx`’s `txid` is included in
+    ///   `header.merkle_root`.
+    /// - If `witness_tx` is `Some`, this checks that the coinbase transaction (the “base_tx”) is
+    ///   correctly included in `header.merkle_root`, and that the coinbase commits to the witness
+    ///   transaction’s `wtxid` in its witness Merkle root.
+    pub(crate) fn verify(&self, header: Header) -> bool {
+        // First, verify that the `base_tx` is in the Merkle tree given by `header.merkle_root`.
+        let merkle_root: Buf32 = header.merkle_root.to_byte_array().into();
+        if !self.base_tx.verify(merkle_root) {
+            return false;
+        }
+
+        match &self.witness_tx {
+            Some(witness) => {
+                let coinbase = self.base_tx.tx.as_ref();
+                // The base transaction must indeed be a coinbase if we are committing
+                // to witness data.
+                if !coinbase.is_coinbase() {
+                    return false;
+                }
+
+                // Compute the witness Merkle root for the transaction in question.
+                let L1TxWithIdProof { tx, proof } = witness;
+                let mut witness_root = proof.compute_root(tx.as_ref()).as_bytes().to_vec();
+
+                // The coinbase input’s witness must have exactly one element of length 32,
+                // which should be the “wtxid” commitment.
+                let witness_vec: Vec<_> = coinbase.input[0].witness.iter().collect();
+                if witness_vec.len() != 1 || witness_vec[0].len() != 32 {
+                    return false;
+                }
+
+                // Append the committed data to the `witness_root` bytes.
+                witness_root.extend(witness_vec[0]);
+
+                // Double SHA-256 of the root + data gives us the final commitment.
+                let commitment = sha256d(&witness_root);
+
+                // Check if the coinbase transaction’s witness commitment matches.
+                match witness_commitment_from_coinbase(coinbase) {
+                    Some(root) => commitment == root.to_byte_array().into(),
+                    None => false,
+                }
+            }
+            None => true,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bitcoin::Block;
+
+    use super::L1TxWithProofBundle;
+
+    #[test]
+    fn test_segwit_tx() {
+        let blocks_bytes = include_bytes!("../../../../test-data/blocks.bin");
```

Comment:
```suggestion
        let blocks_bytes = std::fs::read("../../../../test-data/blocks.bin").unwrap();
```
There is no reason this data needs to be in the test binary.